src/wayland-eglsurface.c

```diff
@@ -286,8 +290,22 @@ wlEglSendDamageEvent(WlEglSurface *surface, struct wl_event_queue *queue)
                           surface->dy);
     }
 
-    wl_surface_damage(surface->wlSurface, 0, 0,
-                      surface->width, surface->height);
+    if (n_rects > 0 &&
+        (wl_proxy_get_version((struct wl_proxy *)surface->wlSurface) >=
+         WL_SURFACE_DAMAGE_BUFFER_SINCE_VERSION)) {
+        for (i = 0; i < n_rects; i++) {
+            int y = rects[i*4+1] + rects[i*4+3];
+            wl_surface_damage_buffer(surface->wlSurface,
+                                     rects[i*4],
+                                     surface->height - y,
+                                     rects[i*4+2],
+                                     rects[i*4+3]);
```

Comment:
Thanks I think this looks pretty good, just a couple nitpicks:

I think we should have a small comment here mentioning that the coordinate systems are flipped between `eglSwapBuffersWithDamage()` and `wl_surface_damage_buffer()`.  Also naming this variable `y` seems slightly confusing, as technically `rects[i*4+1]` is really the passed-in y position and `int y` here is actually the ending point of the damage rectangle. Maybe naming it something like `max_y` would be more accurate?

We have a lot of `i*4` here, I think it might be more readable if we created a `rect_offset` variable to hold this? i.e. `rects[rect_offset+2]`.

Reply:
> We have a lot of i*4 here, I think it might be more readable if we created a rect_offset

For that, it would probably be more readable to just declare a second pointer so that we can index from zero:
```C
const EGLint *rect = &rects[i * 4];
int inv_y = surface->height - rect[1] + rect[3];
wl_surface_damage_buffer(surface->wlSurface, rect[0], inv_y, rect[2], rect[3]);
```

Reply:
Sounds good. I applied the suggestion by @kbrenneman and added a small comment about the Y being flipped.